plugin/pkg/admission/certificates/signing/admission.go

```diff
@@ -0,0 +1,170 @@
+	if !ok {
+		return admission.NewForbidden(a, fmt.Errorf("expected type CertificateSigningRequest, got: %T", a.GetObject()))
+	}
+
+	// only run if the status.certificate field has been changed
+	if reflect.DeepEqual(oldCSR.Status.Certificate, csr.Status.Certificate) {
+		return nil
+	}
+
+	// It is safe for us to dereference signerName because the defaulting logic will run before this admission
+	// controller is executed, meaning we know it is set even if an older client that is not aware of the signerName
+	// field is submitting the signed certificate.
+	signerName := *csr.Spec.SignerName
+
+	if isAuthorizedForPolicy(ctx, a.GetUserInfo(), signerName, p.authz) {
+		return nil
+	}
+
+	// we didn't validate against any provider, reject the pod and give the errors for each attempt
+	klog.V(4).Infof("user not permitted to sign CertificateSigningRequest %q with signerName %q", csr.Name, signerName)
+	return admission.NewForbidden(a, fmt.Errorf("user not permitted to sign requests with signerName %q", signerName))
+}
+
+// isAuthorizedForPolicy returns true if info is authorized to perform the "sign" verb on the synthetic
+// 'signers' resource with the given signerName.
+// It will also check if the user has permission to perform the "sign" verb against the domain portion of
+// the signerName + /*, i.e. `kubernetes.io/*` iff the user doesn't have explicit permission on the named signer.
+func isAuthorizedForPolicy(ctx context.Context, info user.Info, signerName string, authz authorizer.Authorizer) bool {
+	if info == nil {
+		return false
+	}
+
+	// First check if the user has explicit permission to approve for the given signerName.
+	attr := buildAttributes(info, signerName)
+	decision, reason, err := authz.Authorize(ctx, attr)
+	if err != nil {
+		klog.V(5).Infof("cannot authorize for policy: %v,%v", reason, err)
+		return false
+	}
+	if decision == authorizer.DecisionAllow {
+		return true
+	}
+
+	// If not, check if the user has wildcard permissions to approve for the domain portion of the signerName, e.g.
+	// 'kubernetes.io/*'.
+	attr = buildWildcardAttributes(info, signerName)
+	decision, reason, err = authz.Authorize(ctx, attr)
+	if err != nil {
+		klog.V(5).Infof("cannot authorize for policy: %v,%v", reason, err)
+		return false
+	}
+	if decision == authorizer.DecisionAllow {
+		return true
+	}
+
+	return false
+}
+
+// buildAttributes builds an attributes record for a SAR based on the user info and policy.
+func buildAttributes(info user.Info, signerName string) authorizer.Attributes {
+	return buildAttributeRecord(info, signerName)
+}
+
+func buildWildcardAttributes(info user.Info, signerName string) authorizer.Attributes {
+	parts := strings.Split(signerName, "/")
+	domain := parts[0]
+	return buildAttributeRecord(info, domain+"/*")
+}
+
+func buildAttributeRecord(info user.Info, name string) authorizer.Attributes {
+	return authorizer.AttributesRecord{
+		User:            info,
+		Verb:            "sign",
```

Comment:
I think they are included, but hard to spot :eyes:

```
There are congruent check for providing a signature that use the verb=="sign" instead of "approve" above.
```

https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190607-certificates-api.md#limiting-approval-and-signer-powers-for-certain-signers

Reply:
That said, maybe we want to adjust the wording here as it's not that clear unless you've seen the KEP evolve...

Reply:
Ah guess I missed it 😄